### ISSUE TITLE
Commas in cookie value not parsed correctly on Android

### ISSUE THREAD

**Issue by uen** [facebook/react-native#29916]
## Description
tl;dr Commas are treated as cookie deliminators on Android only

Commas (,) in cookie values are not allowed as part of the [cookie specification](https://curl.haxx.se/rfc/cookie_spec.html) but they're still allowed on all major browsers and through iOS's networking APIs. However in Android okhttp (specifically okhttp3.JavaNetCookieJar) [commas are treated as cookie delimiters](https://github.com/square/okhttp/blob/master/okhttp-urlconnection/src/main/kotlin/okhttp3/JavaNetCookieJar.kt#L81) in the same way that semicolons are. This behaviour isn't mentioned anywhere and is inconsistent with iOS/Web.

I have found an existing [issue](https://github.com/square/okhttp/issues/4636) on okhttp which suggests the regular CookieJar doesn't have this limitation and should be used instead of JavaNetCookieJar. I can also confirm that removing the comma [here](https://github.com/square/okhttp/blob/master/okhttp-urlconnection/src/main/kotlin/okhttp3/JavaNetCookieJar.kt#L81) fixes the issue. React Native sets the okhttp cookie jar to JavaNetCookieJar [here](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L184).

I'm working with an external service which I can't control so changing the cookie isn't an option. Because {redirect: "manual"} doesn't work in React Native iOS/Android, it's impossible to override this behaviour if the cookie is set in a 302 response and required for subsequent responses.

## React Native version:
```
System:
    OS: Linux 5.3 Linux Mint 19.1 (Tessa)
    CPU: (4) x64 Intel(R) Core(TM) m3-6Y30 CPU @ 0.90GHz
    Memory: 1.50 GB / 7.67 GB
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 10.15.0 - /usr/local/bin/node
    Yarn: Not Found
    npm: 6.4.1 - /usr/local/bin/npm
    Watchman: Not Found
  SDKs:
    Android SDK:
      Android NDK: 17.2.4988734
  IDEs:
    Android Studio: Not Found
  Languages:
    Java: 11.0.3 - /usr/bin/javac
    Python: 2.7.17 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: ~16.11.0 => 16.11.0 
    react-native: github:facebook/react-native#0.63-stable => 0.63.2 
  npmGlobalPackages:
```

## Steps To Reproduce and expected results

If a fetch request is made to a server which responds with the following header:

```
Set-Cookie: TestCookie=a:hello,b:goodbye
```
On every platform apart from Android, the following Cookie header is sent in future requests:
```
Cookie: TestCookie=a:hello,b:goodbye;
```

but on Android the cookie is parsed incorrectly and the following header is sent (2 cookies):
```
Cookie: TestCookie=a:hello;b:goodbye=;
```
## Snack, code example, screenshot, or link to a repository:
To easily reproduce this, run this express project [https://gist.github.com/uen/f8bbded0fc10fd2d3910d388cd3fea6e](https://gist.github.com/uen/f8bbded0fc10fd2d3910d388cd3fea6e) and load snack [https://snack.expo.io/8y09x_Bc7](https://snack.expo.io/8y09x_Bc7) on a real Android and iOS device (and web). The Android cookie will not be correct as described above

**Comment by uen** [comment]
Are there any plans to change this / would a PR changing the cookie jar be accepted?

**Comment by uen** [comment]
Anyone having this issue, I forked react-native with a fix here: https://github.com/uen/react-native-34/commit/4a054da2169c4f085b6165648a1290b3ea8b3751 - It should be pretty much the same in every other version of RN too.

The fork simply copies the okhttp JavaNetCookieJar class and fixes the issue, and then makes React Native use it. Ideally JavaNetCookieJar needs to be changed to use CookieJar from okhttp, but this may have unintended side effects & probably needs tests updating etc.

I will eventually create a module that replaces the Networking module here: https://github.com/uen/react-native-cookie-fetch so it's more 'drop-in' and won't require a custom RN version.

If you use Expo managed workflow there is no way around this issue and I doubt it will be fixed since there's over 1k issues open in this repo currently

@whyer

**Comment by viktorlarsson** [comment]
I tried creating a combination of your code @uen , and a technique to try to avoid actually changing the RN code. Haven't gotten it to work just yet but:

https://github.com/kolplattformen/skolplattformen/tree/bugfix/cookie-on-android/packages/app/android/app/src/main/java/org/skolplattformen

**Comment by viktorlarsson** [comment]
Sorry, here it is @uen 

https://github.com/kolplattformen/skolplattformen/tree/main/packages/app/android/app/src/main/java/org/skolplattformen/app

**Comment by viktorlarsson** [comment]
Still having issues, hard to debug unfortunately!

**Comment by uen** [comment]
It is still an issue

**Comment by uen** [comment]
It's still an issue

**Comment by uen** [comment]
Hi it's me again from 5 years ago, this problem has REALLY annoyed me and made things mega annoying having to keep building compiled versions of React Native. I came back and decided to fix it, here's a Expo plugin that uses a custom cookie jar and allows commas in cookie values. 

```
const { withMainApplication, createRunOncePlugin } = require('@expo/config-plugins');

const pkg = require('../package.json');

const REQUIRED_IMPORTS = [
  'com.facebook.react.modules.network.CustomClientBuilder',
  'com.facebook.react.modules.network.NetworkingModule',
  'java.io.IOException',
  'java.net.CookieHandler',
  'java.net.CookieManager',
  'java.util.Collections',
  'okhttp3.Cookie',
  'okhttp3.CookieJar',
  'okhttp3.HttpUrl',
  'okhttp3.OkHttpClient',
  'okhttp3.internal.platform.Platform'
];

const SHARED_COOKIE_JAR_PROPERTY = `
  private val sharedCookieJar: CookieJar by lazy {
    CustomJavaNetCookieJar(CookieManager())
  }
`;

const ON_CREATE_NETWORKING_MODULE = `
    NetworkingModule.setCustomClientBuilder(
      object: CustomClientBuilder {
        override fun apply(builder: OkHttpClient.Builder) {
          builder.cookieJar(sharedCookieJar)
        }
      }
    )
`;

const CUSTOM_COOKIE_JAR_CLASS = `
class CustomJavaNetCookieJar(
  private val cookieHandler: CookieHandler,
) : CookieJar {
  override fun saveFromResponse(
    url: HttpUrl,
    cookies: List<Cookie>,
  ) {
    val cookieStrings = mutableListOf<String>()
    for (cookie in cookies) {
      cookieStrings.add(cookie.toString())
    }
    val multimap = mapOf("Set-Cookie" to cookieStrings)
    try {
      cookieHandler.put(url.toUri(), multimap)
    } catch (e: IOException) {
      Platform.get().log("Saving cookies failed for " + url.resolve("/...")!!, 1, e)
    }
  }

  override fun loadForRequest(url: HttpUrl): List<Cookie> {
    val cookieHeaders =
      try {
        cookieHandler.get(url.toUri(), emptyMap<String, List<String>>())
      } catch (e: IOException) {
        Platform.get().log("Loading cookies failed for " + url.resolve("/...")!!,
          1, e)
        return emptyList()
      }

    var cookies: MutableList<Cookie>? = null
    for ((key, value) in cookieHeaders) {
      if (("Cookie".equals(key, ignoreCase = true) || "Cookie2".equals(key, ignoreCase = true)) &&
        value.isNotEmpty()
      ) {
        for (header in value) {
          if (cookies == null) cookies = mutableListOf()
          cookies.addAll(decodeHeaderAsJavaNetCookies(url, header))
        }
      }
    }

    return if (cookies != null) {
      Collections.unmodifiableList(cookies)
    } else {
      emptyList()
    }
  }

  private fun decodeHeaderAsJavaNetCookies(
    url: HttpUrl,
    header: String,
  ): List<Cookie> {
    val result = mutableListOf<Cookie>()
    var pos = 0
    val limit = header.length
    var pairEnd: Int
    while (pos < limit) {
      pairEnd = header.delimiterOffset(";", pos, limit)
      val equalsSign = header.delimiterOffset('=', pos, pairEnd)
      val name = header.trimSubstring(pos, equalsSign)
      if (name.startsWith("$")) {
        pos = pairEnd + 1
        continue
      }

      var value =
        if (equalsSign < pairEnd) {
          header.trimSubstring(equalsSign + 1, pairEnd)
        } else {
          ""
        }

      if (value.startsWith("\"") && value.endsWith("\"") && value.length >= 2) {
        value = value.substring(1, value.length - 1)
      }

      result.add(
        Cookie
          .Builder()
          .name(name)
          .value(value)
          .domain(url.host)
          .build(),
      )
      pos = pairEnd + 1
    }
    return result
  }
}

private fun String.delimiterOffset(
  delimiter: String,
  startIndex: Int = 0,
  endIndex: Int = length
): Int {
  for (i in startIndex until endIndex) {
    if (indexOf(delimiter, i) == i) return i
  }
  return endIndex
}

private fun String.delimiterOffset(
  delimiter: Char,
  startIndex: Int = 0,
  endIndex: Int = length
): Int {
  for (i in startIndex until endIndex) {
    if (this[i] == delimiter) return i
  }
  return endIndex
}

private fun String.trimSubstring(startIndex: Int = 0, endIndex: Int = length): String {
  var a = startIndex
  while (a < endIndex && this[a].isWhitespace()) {
    a++
  }
  var b = endIndex
  while (b > a && this[b - 1].isWhitespace()) {
    b--
  }
  return substring(a, b)
}
`;

  
const withSharedCookieJar = (config) => {
  return withMainApplication(config, async (config) => {
    let mainApplicationContent = config.modResults.contents;

    const splitContents = config.modResults.contents.split('\n');
    const importItems = REQUIRED_IMPORTS.map((importItem) => `import ${importItem}`);
    for(const importItem of importItems) {
      const importItemLine = splitContents.findIndex((line) => line.includes(importItem));
      if(importItemLine === -1) {
        splitContents.splice(1, 0, importItem);
      }
    }

    const classDeclaration = 'class MainApplication : Application()';
    if (!mainApplicationContent.includes(SHARED_COOKIE_JAR_PROPERTY.trim())) {
      const classDeclarationLine = splitContents.findIndex((line) => line.includes(classDeclaration));
      splitContents.splice(classDeclarationLine + 1, 0, SHARED_COOKIE_JAR_PROPERTY);
    }
    
    const superOnCreate = 'super.onCreate()';
    if (!mainApplicationContent.includes('NetworkingModule.setCustomClientBuilder')) {
      const superOnCreateLine = splitContents.findIndex((line) => line.includes(superOnCreate));
      splitContents.splice(superOnCreateLine + 1, 0, ON_CREATE_NETWORKING_MODULE);
    }

    const customCookieJarClassLine = splitContents.findIndex((line) => line.includes('class CustomJavaNetCookieJar'));
    if (customCookieJarClassLine === -1) {
      splitContents.splice(splitContents.length - 1, 0, CUSTOM_COOKIE_JAR_CLASS);
    }

    config.modResults.contents = splitContents.join('\n');
    return config;
  });
};

module.exports = createRunOncePlugin(withSharedCookieJar, pkg.name, pkg.version);
```
